File: infullmvp-basetest/src/main/java/com/infullmobile/android/infullmvp/basetest/InFullMvpDialogFragmentBaseTest.java
package com.infullmobile.android.infullmvp.basetest;

import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;

import org.junit.After;
import org.junit.Before;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.shadows.support.v4.SupportFragmentController;

public abstract class InFullMvpDialogFragmentBaseTest<T extends DialogFragment> {

    private SupportFragmentController<T> fragmentController;
    private T fragment;

    @Before
    public void setUp() {
        fragmentController = SupportFragmentController.of(provideFragment());
        fragment = fragmentController.create().get();
        substituteModules(fragment);
        fragmentController.start().resume().visible();
    }

    @After
    public void tearDown() {
        fragmentController.pause().stop().destroy();
    }

    @NonNull
    protected String getString(int stringResourceId) {
        return RuntimeEnvironment.application.getResources().getString(stringResourceId);
    }

    @NonNull
    protected abstract T provideFragment();


    public T getTestedFragment() {
        return fragment;
    }

    public void substituteModules(T fragment) {
        /* NO OP */
    }
}

File: infullmvp-kotlin-basetest/src/main/java/com/infullmobile/android/infullmvp/basetest/InFullMvpDialogFragmentBaseTest.kt
package com.infullmobile.android.infullmvp.basetest

import com.infullmobile.android.infullmvp.InFullMvpDialogFragment
import com.infullmobile.android.infullmvp.PresentedDialogView
import com.infullmobile.android.infullmvp.Presenter
import org.junit.After

import org.junit.Before
import org.robolectric.RuntimeEnvironment
import org.robolectric.shadows.support.v4.SupportFragmentController

abstract class InFullMvpDialogFragmentBaseTest<T : InFullMvpDialogFragment<*, *>> {

    lateinit var testedFragment: T
        private set

    val testedPresenter: Presenter<*>
        get() = testedFragment.presenter
    val testedView: PresentedDialogView<*>
        get() = testedFragment.presentedView

    private lateinit var fragmentController: SupportFragmentController<T>

    @Before
    open fun setUp() {
        fragmentController = SupportFragmentController.of(provideFragment())
        testedFragment = fragmentController.create().get()
        substituteModules(testedFragment)
        fragmentController.start().resume().visible()
    }

    @After
    fun tearDown() {
        fragmentController.pause().stop().destroy()
    }

    protected fun getString(stringResourceId: Int): String {
        return RuntimeEnvironment.application.resources.getString(stringResourceId)
    }

    protected abstract fun provideFragment(): T

    open fun substituteModules(fragment: T) {
        /* NO OP */
    }
}
